[쇼핑 장바구니 3, 4단계] 페토 미션 제출합니다.

chanho0908 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

### 제이든 안녕하세요 ! 

미션을 진행하다보니 디엠으로 제이든과 이야기 했던 제 이번 미션의 목표를 지키지 못한 것 같습니다 🥲

혹여 오버엔지니어링에 일 수도 있다고 생각되거나 과하다고 생각될 부분들도 있을 것 같아요

그런 부분들은 비동기 콜백으로 인한 타이밍 이슈를 해결하고자 고민하고 나온 내용들입니다 ! ☺️

아이디로 상품을 조회하다보니 여러 Repository를 조합 해야 했고, 이로인해 코드가 굉장히 지저분해지고 
엄청난 콜백지옥에 빠지게 되었습니다 

그래서 이를 최대한 가독성있고 재활용하고자 Repository들을 조합해 도메인을 만드는 Loader라는 클래스들을 만들게 되었습니다

최대한 비동기 콜백 관련 문제들을 해결하고자 다양한 시도들을 해봤는데 이 시도들이 과연 이해하기 쉬운 코드가 될지는 잘 모르겠습니다

그럼 이번 미션도 잘 부탁드립니다 !  ☺️

## 참고 자료


## 스크린샷

https://github.com/user-attachments/assets/b713ff63-1281-4598-bb46-d83a11795e8e

https://github.com/user-attachments/assets/e1c1265a-c407-44aa-9aa3-4eb1353ab640

chanho0908 committed:
feat : 상품 목록, 수량 조절 버튼에 필요한 UI 요소 추가

chanho0908 committed:
refactor : ProductStorage 싱글톤 객체로 변경

chanho0908 committed:
feat : 상품 수량 클래스 추가

chanho0908 committed:
refactor : 장바구니 상품 조회 시 Nullable 반환 타입으로 수정

chanho0908 committed:
feat : 상품 수량 정보 추가 및 페이지네이션 개선

- 상품 데이터에 수량(quantity) 정보를 추가했습니다.
- 상품 목록 조회 시 시작 인덱스가 유효하지 않을 경우 빈 목록을 반환하도록 페이지네이션 로직을 수정했습니다.

chanho0908 committed:
feat : 상품 목록 수량 조절 기능 추가

chanho0908 committed:
feat : MainViewModel에 CartRepository 의존성 주입

chanho0908 committed:
feat : 상품 목록 수량 조절 기능 추가

- 상품 목록에서 상품 수량을 조절할 수 있는 UI 및 기능 추가
- `ProductRvItems`에 `quantity`, `quantityVisible` 추가
- `ProductAdapter`에서 아이템 변경 시 `notifyItemChanged` 호출 로직 수정
- `MainViewModel`에 수량 증가/감소 함수 추가 및 장바구니 연동
- 관련 테스트 코드 수정 및 추가

chanho0908 committed:
docs : 상품 목록의 장바구니 기능 추가

chanho0908 commented:
@coderabbitai pause

chanho0908 committed:
refactor : ProductResult 이름 ProductSinglePage 로 변경

chanho0908 committed:
refactor : `CartResult` 클래스명을 `CartSinglePage`로 변경

chanho0908 committed:
feat : 상품 재고 수정 기능 구현

chanho0908 committed:
feat : Quantity 초과 여부 확인 기능 추가

chanho0908 committed:
refactor : 상품 재고 검증 로직 추가 및 ProductState 위치 변경

chanho0908 committed:
refactor : ProductUiState 클래스 패키지 이동 및 로직 변경

- ProductUiState 클래스를 `woowacourse.shopping.view.main.vm`에서 `woowacourse.shopping.view.main.vm.state`로 이동했습니다.
- 수량 증가/감소 관련 로직을 ProductState로 이전하고, ProductUiState는 해당 ProductState를 교체하도록 수정했습니다.
- 장바구니에 상품 추가 가능 여부, 수량 증가 가능 여부를 확인하는 로직을 추가했습니다.

chanho0908 committed:
refactor : 장바구니에 담긴 상품의 수량 증감 로직 변경

- 상품의 수량을 변경하는 로직을 ViewModel로 이동
- 수량 변경 시 발생하는 이벤트를 정의하여 View에서 처리하도록 수정
- 관련 테스트 코드 수정

chanho0908 committed:
refactor : 상품 수량 수정시 Quantity 객체 받도록 변경

chanho0908 committed:
refactor : 장바구니 상품 id 삭제 및 quantity 필드 위치 변경

- Cart 클래스에서 `id` 필드를 삭제합니다.
- `quantity` 필드의 위치를 `productId` 앞으로 이동합니다.
- `CartRepository.insert()` 메서드가 `Cart`를 생성할 때 `quantity`를 1로 초기화하도록 수정합니다.
- `CartRepository.delete()` 메서드를 `modifyQuantity()`로 변경하여 상품 수량 변경 기능을 제공합니다.

chanho0908 committed:
refactor : 장바구니 수량 수정 로직 변경

- CartStorage에서 Cart의 Quantity를 수정할 수 있도록 변경
- 수량이 0이 되면 장바구니에서 해당 상품을 삭제
- CartStorage 내부 변수명 및 반환 타입 변경

chanho0908 committed:
feat : 상품 목록에서 상품 개수 감소 기능 추가

chanho0908 committed:
feat : 상품 목록에서 상품 개수 감소 기능 추가

chanho0908 committed:
feat : 상품 목록에서 상품 개수 감소 기능 추가

chanho0908 committed:
feat : Toast 확장 함수 추가

chanho0908 committed:
feat : Toast 확장 함수 추가

chanho0908 committed:
feat : 상품 재고 초과시 토스트 메시지 기능 구현

chanho0908 committed:
feat : 상품 재고 초과시 토스트 메시지 기능 구현

chanho0908 committed:
refactor: 수량 변경 UI 레이아웃 분리

- `item_product.xml`의 수량 변경 관련 UI를 `item_quantity_changer.xml`로 분리
- `item_product.xml`에 `item_quantity_changer.xml`을 include하여 사용
- `item_product.xml`의 이벤트 핸들러 이름 변경 및 새로운 이벤트 핸들러 추가

chanho0908 committed:
refactor: 수량 변경 UI 레이아웃 분리

- `item_product.xml`의 수량 변경 관련 UI를 `item_quantity_changer.xml`로 분리
- `item_product.xml`에 `item_quantity_changer.xml`을 include하여 사용
- `item_product.xml`의 이벤트 핸들러 이름 변경 및 새로운 이벤트 핸들러 추가

chanho0908 committed:
refactor : 상품 목록 어댑터 이벤트 핸들러 인터페이스 구조 개선

- `ProductsAdapterEventHandler` 인터페이스를 각 뷰홀더에서 사용하는 핸들러 인터페이스로 분리
- 공통으로 사용되는 장바구니 수량 조절 관련 핸들러를 `CartQuantityHandler`로 분리

chanho0908 committed:
refactor : 상품 목록 어댑터 이벤트 핸들러 인터페이스 구조 개선

- `ProductsAdapterEventHandler` 인터페이스를 각 뷰홀더에서 사용하는 핸들러 인터페이스로 분리
- 공통으로 사용되는 장바구니 수량 조절 관련 핸들러를 `CartQuantityHandler`로 분리

chanho0908 committed:
feat : 장바구니 수량 표시 기능 추가

- 툴바에 장바구니 아이콘과 수량 표시 뱃지 추가
- `badge_background.xml` 파일 추가하여 뱃지 배경 스타일 정의

chanho0908 committed:
feat : 장바구니 수량 표시 기능 추가

- 툴바에 장바구니 아이콘과 수량 표시 뱃지 추가
- `badge_background.xml` 파일 추가하여 뱃지 배경 스타일 정의

chanho0908 committed:
feat : 장바구니 수량 표시 기능 추가

- 툴바에 장바구니 아이콘과 수량 표시 뱃지 추가
- `badge_background.xml` 파일 추가하여 뱃지 배경 스타일 정의

chanho0908 committed:
feat : 장바구니 수량 표시 기능 추가

- 툴바에 장바구니 아이콘과 수량 표시 뱃지 추가
- `badge_background.xml` 파일 추가하여 뱃지 배경 스타일 정의

chanho0908 committed:
feat : 상품 목록에서 상품 수량 변경 UI 배경 추가

상품 목록 아이템 레이아웃(`item_product.xml`)의 수량 변경 부분에 배경(`@drawable/product_quantity_background`)을 추가했습니다.
기존에 `item_quantity_changer.xml`에 있던 배경은 제거하여 중복 적용을 방지했습니다.

chanho0908 committed:
feat : 상품 목록에서 상품 수량 변경 UI 배경 추가

상품 목록 아이템 레이아웃(`item_product.xml`)의 수량 변경 부분에 배경(`@drawable/product_quantity_background`)을 추가했습니다.
기존에 `item_quantity_changer.xml`에 있던 배경은 제거하여 중복 적용을 방지했습니다.

chanho0908 committed:
refactor : ProductRvItems.ProductItem가 ProductState를 직접 참조하도록 수정

- ProductRvItems.ProductItem 내부의 item, quantity, quantityVisible 속성을 ProductState로 대체
- 이에 따라 ProductViewHolder, ProductAdapter, item_product.xml, item_quantity_changer.xml 수정
- 불필요해진 ProductState.toProductRvItems() 확장 함수 제거
- ProductState에 cartQuantityValue 속성 추가

chanho0908 committed:
refactor : ProductRvItems.ProductItem가 ProductState를 직접 참조하도록 수정

- ProductRvItems.ProductItem 내부의 item, quantity, quantityVisible 속성을 ProductState로 대체
- 이에 따라 ProductViewHolder, ProductAdapter, item_product.xml, item_quantity_changer.xml 수정
- 불필요해진 ProductState.toProductRvItems() 확장 함수 제거
- ProductState에 cartQuantityValue 속성 추가

chanho0908 committed:
refactor : 상품 UI 상태 수정 메서드 인자명 변경

chanho0908 committed:
refactor : 상품 UI 상태 수정 메서드 인자명 변경

chanho0908 committed:
refactor : 장바구니에 없는 상품 조회 시, 빈 Cart 객체 반환

chanho0908 committed:
feat : 장바구니 수량 추가 아이콘 벡터 드로어블 변경

chanho0908 committed:
refactor: ProductState에서 quantityVisible 삭제

chanho0908 committed:
refactor : MainActivity의 바인딩 설정 함수 분리

chanho0908 committed:
refactor : 상품 상세 화면 데이터 바인딩 로직 함수 분리

chanho0908 committed:
feat : 상품 상세 화면 수량 조절 기능 추가

- 상품 상세 화면에서 상품의 수량을 조절할 수 있도록 기능을 추가
- 수량 조절 시 최소/최대 수량 제한에 대한 Toast 메시지 표시
- ViewModel과 UiEvent를 통해 수량 변경 및 UI 업데이트 처리

chanho0908 committed:
refactor : 장바구니 수량 변경 시 상품 재고 수량 동기화 수정

- 상품 수량 변경 로직 수정 및 UI 상태 반영
- Product 클래스의 quantity 접근 제어자 변경 (private -> val)
- 테스트 코드 수정

chanho0908 committed:
Merge branch 'chanho0908' of https://github.com/chanho0908/android-shopping-cart into step3

# Conflicts:
#	app/src/main/java/woowacourse/shopping/view/main/MainActivity.kt
#	app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt
#	app/src/main/java/woowacourse/shopping/view/main/vm/state/IncreaseState.kt
#	app/src/main/java/woowacourse/shopping/view/main/vm/state/ProductState.kt
#	app/src/main/java/woowacourse/shopping/view/main/vm/state/ProductUiState.kt
#	app/src/main/res/values/strings.xml
#	app/src/test/java/woowacourse/shopping/view/ProductStateTest.kt

chanho0908 committed:
refactor : 장바구니 수량 변경 시 상품 재고 수량 동기화 수정

- 상품 수량 변경 로직 수정 및 UI 상태 반영
- Product 클래스의 quantity 접근 제어자 변경 (private -> val)
- 테스트 코드 수정

chanho0908 committed:
fix : 상품 상세 화면 진입시 수량 1 증가 오류 수정

chanho0908 committed:
feat : 장바구니 상품 추가 시 수량 지정 기능 추가

- CartRepository의 insert 메서드에 수량(quantity) 파라미터 추가
- CartRepositoryImpl의 get 메서드가 Cart?를 반환하도록 수정
- CartRepositoryImpl의 insert 메서드에 quantity 파라미터 반영

chanho0908 committed:
feat : 상세 화면 장바구니 담기 기능 추가

- 상품 상세 화면에서 '장바구니 담기' 버튼 클릭 시, 해당 상품을 장바구니에 추가하거나 수량을 업데이트하는 기능 구현
- 이미 장바구니에 있는 상품인 경우 수량을 업데이트하고, 없는 경우 새로 추가
- 장바구니 담기 완료 후 장바구니 화면으로 이동하는 이벤트 발생
- 관련 테스트 코드 추가 및 수정

chanho0908 committed:
refactor : 장바구니 상품 수량 조절 UI 구조 변경

- `item_product.xml`에서 수량 조절 레이아웃(`item_quantity_changer.xml`)을 직접 포함하도록 수정
- `item_quantity_changer.xml`에 배경 및 테두리 스타일 적용

chanho0908 committed:
refactor : activity_detail.xml 코드 정리

- LinearLayout 제거 및 include 태그 직접 사용
- Button 속성 순서 변경

chanho0908 committed:
refactor : 장바구니 화면 아이템 UI 및 이벤트 핸들러 구조 변경

- `ItemCartBinding`에서 `CartState` 대신 `ProductState` 사용
- `ItemCartBinding`에 `CartQuantityHandler` 추가
- 장바구니 상품 가격 계산 로직 변경 (단가 * 수량)
- `CartAdapterEventHandler` 인터페이스 삭제 및 `CartViewHolder.Handler`로 통합
- `CartViewHolder.Handler` 인터페이스 내 `onClickDeleteItem` 메서드 인자명 변경 (cartId -> id)

chanho0908 committed:
feat : 장바구니 상품 삭제 기능 재추가

chanho0908 committed:
refactor : Product의 수량 증가 가능 여부 확인 로직 수정

- Product의 `canIncrease` 메서드가 Quantity 객체를 받도록 수정
- ProductState에서 `increaseCartQuantity` 호출 시 Quantity 객체 생성하도록 변경
- Quantity의 `isExceeded` 메서드가 Quantity 객체를 받도록 수정
- 관련 테스트 코드 수정

chanho0908 committed:
refactor : 장바구니 수량 증가 시 상품 재고 계산 로직 수정

- 상품 재고 계산 시 증가된 장바구니 수량을 사용하도록 수정

chanho0908 committed:
refactor : ProductRepository 및 관련 클래스에서 수량 변경 로직 제거

chanho0908 committed:
refactor : `Paging` 클래스의 `resetToLastPageIfEmpty` 메서드 인자 타입 변경 및 관련 테스트 코드 수정

chanho0908 committed:
feat : 장바구니 목록 수량 조절 기능 추가

- CartAdapter 및 CartViewHolder에 CartQuantityHandler를 전달하여 수량 조절 이벤트 처리
- item_cart.xml 레이아웃에 item_quantity_changer.xml 레이아웃을 include하여 수량 조절 UI 추가

chanho0908 committed:
feat : 장바구니 상품 수량 조절 기능 추가

- 장바구니 목록 화면에서 상품 수량 증감 기능 추가
- 수량 변경 시 재고 및 최대 구매 가능 수량 확인
- 수량 변경 불가 시 Toast 메시지 표시
- `CartState` 삭제 및 `ProductState` 재사용
- `CartUiEvent` 추가 및 관련 로직 `CartViewModel`에 구현
- 테스트 코드 수정

chanho0908 committed:
refactor : MainUiEvent, MainState 관련 클래스 패키지 이동

chanho0908 committed:
feat : CartSavingState에 Boolean 값으로부터 객체 생성하는 정적 팩토리 메서드 추가

chanho0908 committed:
refactor : 코드 가독성 개선 및 변수명 명확화

- CartViewModel에서 ProductState 생성 로직의 람다 인자명 변경 (it -> cart)
- ProductState의 isSaveInCart() 메서드를 CartSavingState.of()를 사용하도록 수정
- CartUiState의 isAddedProduct() 메서드 내부에 지역 변수 추가하여 가독성 향상

chanho0908 committed:
test : 상품 목록 화면 UI 테스트 추가

chanho0908 committed:
test : 상품 상세 화면 UI 및 기능 테스트 코드 추가

- 상품 상세 화면이 정상적으로 표시되는지 확인하는 테스트 추가
- 수량 증가/감소 시 가격이 올바르게 변경되는지 확인하는 테스트 추가
- 테스트를 위해 ProductStorage의 상품 가격 정보 수정

chanho0908 committed:
feat : HTTP 통신을 위한 usesCleartextTraffic 설정 추가

- AndroidManifest.xml에 `android:usesCleartextTraffic="true"` 속성을 추가하여 HTTP 통신을 허용

chanho0908 committed:
build : 네트워크 라이브러리 의존성 추가

chanho0908 committed:
feat : 상품 관련 DTO 및 ProductService 인터페이스 추가

- 상품 정보를 담는 ProductDto 추가
- 페이징된 상품 정보를 담는 ProductPageDto 추가
- 상품 조회 및 페이징 기능을 제공하는 ProductService 인터페이스 정의

chanho0908 committed:
feat : 모킹 서버 및 테스트 코드 구현

- MockingServer 클래스에 상품 조회 및 페이지별 상품 목록 조회 기능 구현
- MockProductServiceTest 클래스에 특정 ID 상품 조회 및 첫 페이지 상품 목록 조회 테스트 코드 작성

chanho0908 committed:
refactor : 상품 저장소 DTO 적용

chanho0908 committed:
feat : 상품 데이터 비동기 로드 기능 추가

- ProductRepository 및 ProductRepositoryImpl에 비동기 데이터 로딩을 위한 콜백 기반 메서드 추가
- ProductDatasource를 도입하여 네트워크 서비스로부터 상품 데이터 호출
- 테스트 코드에서 MockK를 사용하여 비동기 로직 검증

chanho0908 committed:
refactor : ProductRepository 로드 콜백으로 변경

- ProductRepository.loadSinglePage, ProductRepository.get 메서드가 콜백을 통해 데이터를 반환하도록 수정
- ViewModel에서 콜백을 통해 UI 상태를 업데이트하도록 변경
- 테스트 코드 수정

chanho0908 committed:
refactor : 상품 데이터 로드 방식 변경 (ProductDatasource 및 MockingServer 도입)

- ProductRepositoryImpl의 생성자 인자를 ProductStorage에서 ProductDatasource로 변경
- MockingServer를 사용하여 ProductDatasource를 초기화하도록 AppContainer 수정

chanho0908 committed:
feat : Room 라이브러리 추가

- Room 라이브러리 버전 관리 및 의존성 추가

chanho0908 committed:
feat: Room DB 및 CartDao 추가

# Conflicts:
#	app/src/androidTest/java/woowacourse/shopping/data/db/CartDaoTest.kt
#	app/src/main/java/woowacourse/shopping/data/db/dao/CartDao.kt

chanho0908 committed:
feat: Room DB 및 CartDao 추가

chanho0908 committed:
fix: 상품 이미지 경로명 수정

chanho0908 committed:
feat : room Cart Dao, Database 추가

# Conflicts:
#	app/src/androidTest/java/woowacourse/shopping/fixture/CartEntityFixture.kt

chanho0908 committed:
feat: Room DB를 AppContainer에 추가

chanho0908 committed:
fix: 페이지네이션 쿼리 OFFSET 수정

chanho0908 committed:
feat: ProductsDataSource 및 관련 테스트 추가

- `ProductService`를 사용하여 상품 데이터를 가져오는 `ProductsDataSource` 클래스 추가
- `getProduct`, `getProducts`, `singlePage` 메서드를 통해 특정 상품, 여러 상품, 페이징된 상품 목록을 가져오는 기능 구현
- `ProductsDataSource`의 기능들을 검증하는 단위 테스트 추가

chanho0908 committed:
feat: 상품 ID 리스트로 상품 목록 조회 기능 추가

- MockProductService에 상품 ID 리스트로 상품 목록을 조회하는 기능 추가
- MockingServer에 해당 기능을 위한 `getProducts` 메서드 구현
- MockProductServiceTest에 관련 테스트 코드 추가

chanho0908 committed:
refactor: ProductRepository 메서드 시그니처 변경 및 테스트 코드 수정

- `get(productId: Long, onResult: (Product) -> Unit)` 메서드를 `get(productId: Long): Product` 로 변경
- `getProducts(productIds: List<Long>, onResult: (List<Product>) -> Unit)` 메서드 추가
- `ProductRepositoryImpl` 및 관련 테스트 코드 수정

chanho0908 committed:
refactor: CartRepository 및 구현체 비동기 처리 및 DataSource 도입

- CartRepository의 모든 메서드를 비동기 콜백 방식으로 변경
- CartRepositoryImpl에 CartDataSource를 주입하여 사용하도록 수정
- CartRepositoryImpl의 각 메서드를 별도 스레드에서 실행하도록 변경
- CartDataSource 클래스 및 FakeCartRepositoryImpl (테스트용) 추가
- CartRepositoryImplTest의 테스트 케이스를 변경된 Repository 및 DataSource에 맞게 수정

chanho0908 committed:
refactor: 장바구니 수량 증가 로직 및 상품 삭제 로직 단순화

- `CartUiState`에서 `isAddedProduct` 메서드 제거
- `CartViewModel`에서 장바구니에 상품이 담겨있는지 여부에 따라 분기하던 로직을 `upsert`를 사용하여 단순화
- 상품 삭제 후 목록을 새로고침하는 로직을 `refresh` 메서드로 분리하여 재사용성 향상

chanho0908 committed:
refactor: 상품 로드 및 장바구니 저장 로직 비동기 처리 변경

- DetailViewModel에서 상품 정보를 불러오는 로직을 백그라운드 스레드에서 실행하도록 수정했습니다.
- 장바구니에 상품을 저장하는 로직에서 기존에 있던 상품인지 확인하는 방식을 콜백 기반으로 변경하고, upsert와 modify 메서드 사용 로직을 조정했습니다.
- 관련 테스트 코드에서 MockK의 relaxed 설정을 추가하고, 변경된 메서드 호출을 반영했습니다.

chanho0908 committed:
refactor: 장바구니 수량 증가/감소 로직 수정 및 테스트 코드 업데이트

- `CartRepository`에 `getCarts`, `upsert`, `delete` 메서드를 추가하고, `MainViewModel`에서 이를 사용하도록 변경했습니다.
- `MainViewModel`의 `increaseCartQuantity`와 `decreaseCartQuantity` 메서드의 로직을 수정하여 `upsert`와 `delete`를 호출하도록 변경했습니다.
- `MainViewModelTest`에서 `cartRepository`의 `getCarts` 메서드에 대한 mock 객체를 설정하도록 수정했습니다.
- `MainViewModelTest`에서 AssertJ의 `Assertions`를 static import 하도록 수정했습니다.

chanho0908 committed:
refactor: ProductDataSource 클래스명 ProductsDataSource로 변경

chanho0908 committed:
refactor: ProductDto 및 ProductPageDto를 ProductEntity 및 ProductPageEntity로 이름 변경

chanho0908 committed:
refactor: 장바구니에 상품 추가 시 upsert 로직으로 변경

- CartDataSource의 modify 함수 로직을 upsert와 동일하게 수정
- 테스트 코드의 상품명 및 가격 데이터 변경 사항 반영

chanho0908 committed:
refactor: CartStorage 제거

chanho0908 committed:
refactor: CartRepository의 delete 메서드 onResult 콜백을 nullable로 변경

CartRepository 및 구현체(CartRepositoryImpl, FakeCartRepositoryImpl)의 delete 메서드에서 onResult 콜백 파라미터를 nullable로 수정했습니다.
이를 통해 콜백이 필요 없는 경우 null을 전달할 수 있도록 하여 유연성을 높였습니다.

chanho0908 committed:
feat: 장바구니 수량 동기화 기능 추가

- MainActivity의 onResume 시점에서 장바구니 수량 동기화 로직 호출
- MainViewModel에 syncCartQuantities 메서드 추가
  - 현재 UI 상태의 상품 ID 목록으로 장바구니 정보 조회
  - 조회된 장바구니 수량으로 UI 상태 업데이트
- MainViewModel 리팩터링:
  - 수량 증가/감소 로직을 별도 메서드로 분리 (handleIncreaseQuantity, handleDecreaseQuantity)
  - _uiState.value 접근을 withUiState 확장 함수로 캡슐화
- ProductUiState에 productIds 프로퍼티 추가

chanho0908 committed:
refactor: `withUiState` 함수를 사용하여 UI 상태 처리 개선

chanho0908 committed:
refactor: 상품 데이터 로드 방식 비동기 처리로 변경

- `ProductRepository`의 `getProduct` 메서드를 콜백 형태로 변경하여 비동기 데이터 로드를 지원합니다.
- `DetailViewModel` 및 `CartViewModel`에서 상품 정보를 비동기적으로 로드하도록 수정합니다.
- 불필요한 `thread` 블록을 제거하고, 비동기 처리는 `ProductRepositoryImpl` 내부에서 담당하도록 변경합니다.

chanho0908 committed:
feat: HistoryDao 및 HistoryEntity 추가

chanho0908 committed:
feat: HistoryDataSource 및 HistoryRepository 구현

chanho0908 committed:
feat: MainViewModel에 HistoryRepository 추가

chanho0908 committed:
feat: 상품 상세 화면 이동 전 최근 본 상품 저장

MainViewModel에 saveHistory 메서드를 추가하여 상품 상세 화면으로 이동하기 전에 해당 상품 ID를 HistoryRepository를 통해 저장하도록 구현했습니다.

chanho0908 committed:
refactor: `CartSavingState` Enum 삭제

- `ProductState`에서 `CartSavingState` 반환 로직 제거
- 관련 테스트 코드 수정

chanho0908 committed:
feat: 구분선 ViewHolder 추가

chanho0908 committed:
feat: 최근 본 상품 ViewHolder 추가

chanho0908 committed:
feat: 최근 본 상품 레이아웃 ViewHolder 추가

chanho0908 committed:
feat: 최근 본 상품 기능 추가

- `ProductRvItems`에 `RecentProductItem` 및 `DividerItem` 추가
- `ProductAdapter`에서 최근 본 상품 목록 표시 로직 구현
- 최근 본 상품 아이템을 위한 `RecentProductAdapter` 및 `HistoryState` 추가

chanho0908 committed:
feat: ProductLoader 및 HistoryLoader 추가

- 상품 목록 및 구매 내역 로딩 로직을 담당하는 Loader 클래스들을 추가했습니다.
- ProductLoader: 특정 페이지의 상품 정보와 장바구니 수량을 함께 로드합니다.
- HistoryLoader: 구매 내역에 해당하는 상품 정보를 로드합니다.

chanho0908 committed:
feat: 최근 본 상품 기능 추가

- MainActivity에 최근 본 상품 표시 기능 추가
- ProductUiState에 최근 본 상품 관련 필드(historyItems) 추가
- MainViewModel에 최근 본 상품 로드 로직 추가 및 ProductLoader, HistoryLoader 적용

chanho0908 committed:
refactor: 장바구니 수량 조회 로직 개선

- `getCarts` 호출 시 반환되는 `List<Cart?>`를 `Map<Long?, Cart>`로 변환하여 상품 ID로 직접 장바구니 정보를 조회하도록 변경했습니다.
- 이를 통해 `products` 리스트와 `carts` 리스트의 순서에 의존하지 않고, ID를 기준으로 정확하게 장바구니 수량을 매칭할 수 있도록 수정했습니다.

chanho0908 committed:
feat: StateHandler 유틸리티 함수 추가

- `withState` 인라인 함수를 추가하여 상태가 null이 아닐 경우에만 주어진 블록을 실행하도록 함

chanho0908 committed:
refactor: ProductUiState 업데이트 로직 변경

- ProductUiState의 각 필드를 개별 MutableLiveData로 분리
- MediatorLiveData를 사용하여 개별 LiveData를 조합하여 ProductUiState를 생성하도록 변경
- 이에 따라 관련 로직 수정

chanho0908 committed:
feat: 최근 본 상품을 선택시 저장 기능 추가

chanho0908 committed:
Refactor: 최근 본 상품 RecyclerView 성능 개선 및 UI 조정

chanho0908 committed:
feat: 마지막으로 본 상품 ID를 상세 화면으로 전달

chanho0908 committed:
feat: 마지막으로 본 상품 ID를 상세 화면으로 전달

- 상세 화면 진입 시 마지막으로 본 상품 ID를 함께 전달하도록 수정
- `DetailUiState`에 `lastSeenProduct` 필드 추가
- `DetailViewModel`에서 `load` 호출 시 `lastSeenProductId`를 받아 `lastSeenProduct` 정보를 로드하고 `DetailUiState`를 업데이트하도록 변경
- 이에 따라 `activity_detail.xml`에서 최근 본 상품 정보를 표시하는 UI 추가
- 관련 테스트 코드 수정

chanho0908 committed:
feat: 마지막으로 본 상품 클릭 시 해당 상품 상세 화면으로 이동 기능 추가

- `DetailViewModel`에 `HistoryRepository` 의존성 추가
- `loadLastSeenProduct` 함수를 추가하여 클릭된 상품 ID를 `HistoryRepository`에 저장하고, `DetailUiEvent.MoveToLastSeenProduct` 이벤트를 발생시켜 상품 상세 화면으로 이동하도록 구현
- `DetailActivity`에서 `MoveToLastSeenProduct` 이벤트를 관찰하여 새 `Intent`로 `DetailActivity`를 다시 시작하고 현재 `Activity`를 종료하도록 처리
- `activity_detail.xml`에서 마지막으로 본 상품 레이아웃에 `onClick` 핸들러를 추가하여 `vm.loadLastSeenProduct` 함수 호출

chanho0908 committed:
test: CartViewModelTest`의 상품 삭제 테스트 로직 변경사항으로 인한 불필요한 테스트 제거

chanho0908 committed:
Refactor: `ProductLoader` 클래스명 `ProductWithCartLoader`로 변경

chanho0908 committed:
refactor: 장바구니 수량 누적 로직 제거

- `CartDataSource` 및 `CartRepository`에서 `modify` (수량 누적) 기능을 삭제했습니다.
- 이제 `upsert`를 사용하여 장바구니에 상품을 추가하거나 수량을 업데이트합니다.

chanho0908 committed:
feat: CartLoader 클래스 추가

- 장바구니에 담긴 상품 목록을 페이지별로 로드하는 `CartLoader` 클래스를 추가했습니다.
- `CartLoader`는 `CartRepository`와 `ProductRepository`를 사용하여 장바구니 정보와 상품 정보를 가져옵니다.
- 페이지 인덱스와 페이지 크기를 입력받아 해당 페이지의 장바구니 상품 목록을 `ProductState` 리스트로 변환하고, 다음 페이지 존재 여부와 함께 콜백 함수로 전달합니다.

chanho0908 commented on app/src/main/java/woowacourse/shopping/view/core/common/StateHandler.kt:
```diff
@@ -0,0 +1,8 @@
+package woowacourse.shopping.view.core.common
+
+inline fun <T> withState(
+    state: T?,
+    block: (T) -> Unit,
+) {
+    state?.let(block)
+}
```
https://github.com/woowacourse/android-shopping-cart/pull/104#discussion_r2090843534

1,2 단계에서 말씀해주신 내용을 고민해보고 `null`에 대한 중복 체크 문제를 해결하기위해 만들었습니다 !

chanho0908 committed:
Refactor: 변수명 변경 (lastWatchedProductId -> lastSeenProductId)

chanho0908 committed:
feat: MediatorLiveData 확장 함수 추가

- `addSourceList` 확장 함수를 추가하여 여러 LiveData를 한 번에 MediatorLiveData에 추가할 수 있도록 개선
- `MainViewModel`에서 `addSourceList`를 사용하도록 수정

chanho0908 commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,33 +1,163 @@
+            addSource(productItems) { combine() }
+            addSource(historyItems) { combine() }
+            addSource(loadState) { combine() }
+        }
```
`Loader class`로 상품 목록 화면에 필요한 데이터들을 초기화해도 타이밍 이슈로 인해 
최근 본 상품에 대한 목록이 늦게 뜨거나 상품 목록이 늦게 로딩되는 문제가 발생했습니다

각각의 뷰홀더에 필요한 아이템을을 한번에 동기화하기 위해 `ProductUiState` 에 필요한 데이터들을
각각의 LiveData로 나누고 이를 한번에 합쳐 동기화하는 식으로 해결해봤습니다 !

chanho0908 commented:
@coderabbitai resume

chanho0908 committed:
feat: DetailUiState에 productPrice와 hasLastSeenProduct 필드 추가

chanho0908 committed:
Refactor: `ProductState`에 상품 가격 및 장바구니 수량 관련 프로퍼티 추가

chanho0908 committed:
refactor: `CartViewModel`의 `withUiState` 함수를 `withState` 유틸리티 함수로 대체

KwonDae commented on app/src/main/java/woowacourse/shopping/view/cart/adapter/CartAdapter.kt:
```diff
@@ -4,13 +4,15 @@ import android.view.LayoutInflater
+    private var items: List<ProductState>,
+    private val handler: Handler,
+    private val cartQuantityHandler: CartQuantityHandler,
 ) : RecyclerView.Adapter<CartViewHolder>() {
```
이번 단계에서는 LIstAdapter를 활용해보는건 어떨까요?
ListAdapter와 어떤 차이점이 있는지도 함께 찾아보셔도 좋습니다

chanho0908 replied:
https://github.com/woowacourse/android-shopping-cart/pull/121/commits/9b2e8845113bc978d8c7bef1468c417cf2c84477

해당 내용 관련 커밋입니다 !

`ListAdapter`는 사용자가 직접 리스트를 다루지 않아도 DiffUtil을 사용해 리스트의 차이를 계산하고 교체해줍니다.

이 마법같은 일이 어떻게 가능할까 찾아보니 내부적으로 `ListUpdateCallback`을 사용해 그 동안 제가 직접 리스트를 계산하던 작업을 계산해주더군요 ! 

![image](https://github.com/user-attachments/assets/079d5445-ee66-4ac1-a17c-4ebc29b7812d)

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/adapter/ProductAdapter.kt:
```diff
@@ -29,14 +26,18 @@ class ProductAdapter(
                     parent,
-                    handler,
+                    handler as ProductViewHolder.Handler,
+                    handler as CartQuantityHandler,
                 )
 
             ProductRvItems.ViewType.VIEW_TYPE_LOAD ->
-                LoadViewHolder(
-                    parent,
-                    handler,
-                )
+                LoadViewHolder(parent, handler)
+
+            ProductRvItems.ViewType.VIEW_TYPE_RECENT_PRODUCT ->
+                RecentProductViewHolder(parent, handler)
+
+            ProductRvItems.ViewType.VIEW_TYPE_DIVIDER ->
+                DividerViewHolder(parent)
```
여러 타입의 List를 표현하기 위해 Adapter 내 Adapter가 존재하게 구성해주셨네요!
ViewType 에 따라 중첩된 RecyclerView를 적용하다 보면 depth가 깊어지고, 구조가 복잡해지는 경우가 많습니다.

Recyclerview 1.2.0 버전부터 추가 된 ConcatAdapter 라는 키워드도 존재하는데요
어떤 차이점이 있는지 찾아보시고, 개선해보셔도 좋습니다 👍

chanho0908 replied:
ConCatAdapter는 사용해보지 않은 개념이라 시도했으나 수정해야할 부분이 너무 많이 발생해
취사 선택했습니다 

방학때라도 시간 나면 꼭 해보겠습니다 !

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,33 +1,162 @@
+            productWithCartLoader(INITIAL_PAGE, PAGE_SIZE) { productStates, hasNextPage ->
+                historyItems.postValue(historyStates)
+                productItems.postValue(productStates)
+                loadState.postValue(LoadState.of(hasNextPage))
```
postValue는 setValue와 어떤 차이점이 있을까요?

chanho0908 replied:
`setValue`는 `MainThread`에서만 사용하도록 강제하고 있습니다.

`setValue` 메서드 내부를 보면 `assertMainThread` 구문에서 메인 스레드가 아닐 경우 에러가 발생합니다 !
<img width="377" alt="image" src="https://github.com/user-attachments/assets/775f08a0-ccd1-446d-9506-abc3c15f419f" />

때문에 메인 스레드가 아닌 다른 워커 스레드에서 라이브 데이터에 값을 set할 수 있는 `postValue`가 존재합니다
다만 `postValue`또한 값을 쓰는 행위 자체는 MainThread에서 하는 것 같군요 !
<img width="590" alt="image" src="https://github.com/user-attachments/assets/e788d1da-6a7f-4a2e-be2c-a0ac8e5aaf7f" />

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,33 +1,162 @@
+                _uiEvent.postValue(MainUiEvent.NavigateToDetail(productId, state.lastSeenProductId))
+            }
+        }
+    }
```
함수명은 saveHistory인데, 실제 Event는 Navigate 입니다.
interface에서 "히스토리를 저장하는구나~ 화면 이동은 어디서하지?" 하고 들어오니 알게되었네요
뭔가 어색하지 않을까요? 🧐

chanho0908 replied:
https://github.com/woowacourse/android-shopping-cart/pull/121/commits/4debdd965526532af1f5e5bfa817829752e4d0c9

해당 내용 관련 커밋입니다 !

말씀해주신 내용과 더불어 크루들과 최근 본 상품 저장 시점에 대해서 논의했는데요 !

네트워크 이슈로 인해 상품을 실제로 확인하지 못했음에도 저장되는 문제를 방지하기 위해, 최근 본 상품은 상세 화면 로딩에 성공한 경우에만 저장되도록 수정했습니다.

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/MainActivity.kt:
```diff
@@ -78,31 +92,65 @@ class MainActivity : AppCompatActivity(), ProductsAdapterEventHandler {
+
+    override fun onClickDecrease(productId: Long) {
+        viewModel.decreaseCartQuantity(productId)
     }
```
이벤트 처리에 대한 구현은 VIewModel 에서 구현할 수 있지 않을까요?
핸들러 구현을 ViewModel 이 담당하는게 어떨지에 대한 의도입니다

chanho0908 replied:
https://github.com/woowacourse/android-shopping-cart/pull/121/commits/d60ed93702af5dc644d7eef845def844e2a0bc06

https://github.com/woowacourse/android-shopping-cart/pull/121/commits/b5ec9333d0388620c0d3534d51ea7745acd78b99

https://github.com/woowacourse/android-shopping-cart/pull/121/commits/858a513b85f502e77aa85244ac27e0f07a816911

해당 내용 관련 커밋입니다 !

관련 인터페이스의 구현을 ViewModel에 익명 객체로 수정했습니다 !

`ViewModel Class`에서 해당 인터페이스들을 implementation 하는 것도  고려했으나

이 경우 다음과 같은 문제가 있다고 생각하여 익명 객체로 만들었습니다 !
1. ViewModel 을 Adapter에 직접 전달
2. ViewModel이 ViewModel의 역할 뿐만 아니라 interface들의 구현체 역할도 하게된다

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/MainActivity.kt:
```diff
@@ -78,31 +92,65 @@ class MainActivity : AppCompatActivity(), ProductsAdapterEventHandler {
+    override fun onResume() {
+        super.onResume()
+        viewModel.syncHistory()
+        viewModel.syncCartQuantities()
```
아마도 리스트화면으로 돌아왔을 때, 장바구니 담기에 대한 싱크를 맞추기 위해 여기에 들어간걸로 보이네요?

그럼 사용자가 백그라운드에서 다시 포그라운드로 올라오거나, Dialog를 마주했다가 Dialog가 사라지면 어떤 영향이 있을까요?
-> 이건 꼭 생각해보세요.

어떻게 개선할 수 있을까요?
[힌트](https://developer.android.com/training/basics/intents/result)

chanho0908 replied:
https://github.com/woowacourse/android-shopping-cart/pull/121/commits/fd60065453e9be792fc2cfa4fbaf0b3494075021

해당 내용 관련 커밋입니다 !

첫 번째 문제로, 

다른 화면에서 메인 액티비티로 돌아왔을 때만 장바구니 수량 동기화를 의도했던 것과는 다르게 액티비티가 처음 생성될 때 생명주기  콜백 함수 호출로 인해 API를 불필요하게 두 번 호출하는 문제를 확인했습니다

두 번째 문제는

`onResume`은 액티비티가 다시 `foreground`로 돌아왔을 때 호출됩니다.

다이얼로그의 경우 같은 화면`window` 위에 잠깐 보여지므로 다시 `resume`이 호출되는데,
이말은 그냥 간단한 다이얼로그를 보여줄 때도 API가 호출되 불필요한 호출이 발생합니다

이렇게 생각지도 못한 문제들이 발생하구 있었군요 😓

상품 목록 화면 외에, 장바구니 수량이 변경되는 경우는 장바구니 화면에 갔을 경우 뿐이라
`ActivityResult CallBack`을 사용해 정확히 장바구니 화면에 갔다 돌아올 경우에만 동기화 하도록 수정햇습니다 !

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/adapter/DividerViewHolder.kt:
```diff
@@ -0,0 +1,12 @@
+import woowacourse.shopping.databinding.ItemDividerBinding
+import woowacourse.shopping.view.core.base.BaseViewHolder
+
+class DividerViewHolder(
+    parent: ViewGroup,
+) : BaseViewHolder<ItemDividerBinding>(
+        ItemDividerBinding.inflate(LayoutInflater.from(parent.context), parent, false),
+    )
```
ConcatAdapter에 이어지는 내용입니다.
ConcatAdapter를 활용하면 이 뷰홀더는 없어질 수 있을까요?

chanho0908 replied:
일정상의 문제로 ConCatAdapter는 적용하지 못했습니다 🥲

KwonDae commented on app/src/main/java/woowacourse/shopping/view/loader/CartLoader.kt:
```diff
@@ -0,0 +1,51 @@
+import woowacourse.shopping.domain.repository.ProductRepository
+import woowacourse.shopping.view.main.state.ProductState
+
+class CartLoader(
+    private val cartRepository: CartRepository,
+    private val productRepository: ProductRepository,
```
Data 송수신을 담당하는 객체로 잘 분리해주셨습니다 👍 
[이거](https://developer.android.com/topic/architecture/domain-layer?hl=ko)는 참고용으로만 봐주세요.
안드로이드 아키텍처에는 이러한 Repository를 어떻게 관리할 수 있을지 가이드해주고 있습니다.

KwonDae commented on app/src/main/java/woowacourse/shopping/domain/cart/Cart.kt:
```diff
@@ -1,8 +1,8 @@
+import woowacourse.shopping.domain.Quantity
 
 data class Cart(
-    val id: Long = UUID.randomUUID().mostSignificantBits,
```
결국 Cart Id는 사라지게 되었군요..!

KwonDae commented on app/src/main/java/woowacourse/shopping/data/network/MockingServer.kt:
```diff
@@ -0,0 +1,82 @@
+class MockingServer : ProductService {
+    private val mockWebServer = MockWebServer()
+    private val client = OkHttpClient()
+    private val gson = Gson()
```
Gson을 선택한 배경에는 어떤게 있었을까요?
또 이외의 직렬화/역직렬화 라이브러리에는 어떤 것이 있고 어떤 차이점이 있을까요?

KwonDae commented on app/src/main/java/woowacourse/shopping/view/detail/DetailActivity.kt:
```diff
@@ -51,13 +68,25 @@ class DetailActivity : AppCompatActivity() {
+
+                is DetailUiEvent.MoveToLastSeenProduct -> {
+                    startActivity(newIntent(this, it.productId))
+                    finish()
```
> 마지막으로 본 상품 페이지에서 뒤로 가기를 하면 상품 목록으로 이동한다.

위 요구사항을 충족시키기 위해 명시적으로 finish를 해주셨네요.
우리는 Intent Flag라는 키워드를 활용할 수 있습니다.

어떤 [Flag](https://developer.android.com/reference/android/content/Intent#flags)를 사용할 수 있을지 고민해보시고, 개선해주세요!

chanho0908 replied:
https://github.com/woowacourse/android-shopping-cart/pull/121/commits/7bac0c42869653236b2ad03ca6cd02d7c792a015

해당 내용 관련 커밋입니다 !

`Intent.FLAG_ACTIVITY_CLEAR_TOP` 플래그를 사용하여 스택의 상단에 있는 동일한 `Activity`를 제거하고 새 `Intent`로 해당 `Activity`를 다시 시작하도록 수정했습니다 !

KwonDae reviewed (changes requested):
페토 안녕하세요!
구조적으로 설계를 잘 해주셨습니다.
완성도가 높게 구현해주셨던 만큼, 디테일한 부분을 주로 코멘트를 남겨드렸어요.
천천히 고민해보시고 요청 부탁드립니다!

chanho0908 commented:
@coderabbitai review

chanho0908 committed:
refactor: `entitiy` 패키지명 오타 수정 `entity`로 변경

chanho0908 committed:
refactor: 장바구니 상품 목록 조회 성능 개선

- `getCartsByProductIds` 함수에서 각 상품 ID에 대해 개별적으로 쿼리하던 방식을 개선하여, 여러 상품 ID를 한 번의 쿼리로 조회하도록 변경했습니다.
- 이를 위해 `CartDao`에 `IN` 절을 사용하는 `cartsByProductIds` 쿼리를 추가했습니다.

chanho0908 committed:
feat: HistoryEntity productId 기본 키 자동 생성 제거 및 DB 버전 업데이트

chanho0908 committed:
Refactor: ProductsDataSource에서 상품 목록 조회 방식 변경

chanho0908 committed:
refactor: ProductStorage의 ProductEntity 변수명 변경

KwonDae commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,33 +1,162 @@
+        productId: Long,
+    ) {
+        when (state) {
+            is IncreaseState.CanIncrease -> {
+                val newState = state.value
+                _uiState.value = uiState.modifyUiState(newState)
+                cartRepository.upsert(productId, newState.cartQuantity)
+            }
```
이거는 어려울 수 있지만 한번 고민해보시길 바랄게요.
현재는 로직으로 내부 상태 업데이트 & DB 업데이트를 각각 해주고 있습니다.

만약, DB 업데이트에서 이슈가 발생해 정상 실행이 되지 않았다면 어떤 일이 발생할까요?

chanho0908 replied:
사용자가 10개의 상품을 추가했을 때 이슈가 발생해 1개의 상품이 추가되지 않았다면,
상태는 10개로 업데이트 된 반면 실제 DB에는 9개가 추가됩니다

이 경우, 사용자가 상품을 주문했을 때 실제로는 9개가 주문되는 문제가 발생할 수 도 있을 것 같군요 😓

실제 서비스였다면 고객 센터에 불이 났을거 같아요 😱

이 경우를 실제 API를 사용하는 환경에 대입해보면

`수량 증가 -> API 호출 -> 성공시 수량 증가` 

이렇게 되는 방식이 가장 안전할 것 같지만 버튼을 우다다다다 누를 경우 api 응답에 걸리는 시간으로 인해
UX적으로 굉장히 불편할 것 같습니다

때문에 Optimistic UI 같은 방어 기법을 사용하는 것도 방법이 될 것 같습니다 !

chanho0908 committed:
refactor: 장바구니 화면 핸들러 구현 위치 변경

chanho0908 committed:
refactor: `CartQuantityHandler` 구현을 DetailActivity에서 DetailViewModel로 이동

chanho0908 committed:
refactor: ProductAdapterEventHandler 구현체를 ViewModel로 이동

chanho0908 committed:
refactor: 상품 상세에서 목록으로 돌아갈 때 데이터 동기화 방식 변경

chanho0908 committed:
refactor: ProductAdapter를 ListAdapter로 변경하여 DiffUtil 사용

chanho0908 committed:
feat: CartAdapter에 DiffUtil 적용

chanho0908 committed:
refactor: 상품 히스토리 저장 방식 변경

- DetailViewModel에서 히스토리를 저장하도록 변경
- HistoryRepository의 saveHistory 메서드에서 콜백 제거
- MainViewModel에서 HistoryRepository 의존성 제거

chanho0908 committed:
refactor: 불필요한 코드 제거

- `ProductUiState`에서 사용되지 않는 `productItemCount` 및 `addItems` 메서드 제거
- `RecentProductAdapter`에서 불필요한 `init` 블록 내의 로그 출력 코드 제거

chanho0908 committed:
refactor: 최근 본 상품 이동 시 DetailActivity 재활용

최근 본 상품으로 이동할 때, 새로운 DetailActivity를 시작하는 대신 기존 Activity를 재활용하도록 변경했습니다.
Intent.FLAG_ACTIVITY_CLEAR_TOP 플래그를 사용하여 스택의 상단에 있는 동일한 Activity를 제거하고 새 Intent로 해당 Activity를 다시 시작합니다.

chanho0908 committed:
refactor: MainViewModel 생성자에서 historyRepository 제거

chanho0908 committed:
refactor: 핸들러 네이밍 수정

`ProductAdapterEventHandler`에서 `EventHandler`로, `CartAdapter.Handler`에서 `EventHandler`로 네이밍을 수정했습니다.
`productAdapterEventHandler`는 `productEventHandler`로, `cartAdapterHandler`는 `cartEventHandler`로 변경했습니다.

chanho0908 reviewed (commented):
안녕하세요 제이든 !

4단계 미션 진행으로 인해 제출이 다소 늦어졌습니다 😖

리뷰주신 내용 중 ConcatAdapter를 꼭 적용해보고 싶어 시도해봤지만
수정해야할 부분이 너무나 많아져 일정상의 이유로 적용하지 못했습니다

4단계가 너무 어려워 처음으로 미션을 미완성으로 제출해버렸어요 🥲

방학에라도 시간을 내서 꼭 적용해보겠습니다 !

KwonDae reviewed (approved):
마지막까지 피드백 반영하시느라 고생많으셨습니다.
시간이 오래걸리더라도 끝까지 시도하는 모습이 인상깊었어요.
이번 미션에서 경험했던 부분들이 조금이라도 도움이 되었으면 좋겠습니다.
마지막까지 응원할게요 💪
